File: jenjin-core/src/main/java/com/binarymonks/jj/Game.java
package com.binarymonks.jj;

import com.badlogic.gdx.ApplicationListener;
import com.binarymonks.jj.backend.Global;


/**
 * This is the root thing to implement and provide to the libgdx frontend.
 * It is the engine for the application lifecycle and gameloop.
 * <p>
 * Do your initialisation stuff in {@link Game#gameOn()}.
 * <p>
 * If you want to hook into the application lifecycle close to the bone you can
 * do it here by overriding the relevant method - be sure to call super though :D
 * Alternatively - to save having to pass the hook down your object graph to where you want it
 * you can register a {@link com.binarymonks.jj.lifecycle.LifeCycleAware} at {@link JJ}
 */
public abstract class Game implements ApplicationListener {

    JJConfig jjConfig = new JJConfig();

    public Game() {
    }

    public Game(JJConfig jjConfig) {
        this.jjConfig = jjConfig;
    }

    @Override
    public void create() {
        JJ.initialise(this.jjConfig);
        gameOn();
    }

    protected abstract void gameOn();

    @Override
    public void render() {
        Global.time.update();
        JJ.assets.update();
        Global.tasks.preloopTasks.update();
        Global.thingWorld.update();
        Global.layerStack.update();
        Global.physics.update();
        Global.tasks.postPhysicsTasks.update();
    }

    @Override
    public void resize(int width, int height) {
        Global.lifecycle.resize(width, height);
    }

    @Override
    public void pause() {
        Global.lifecycle.pause();
    }

    @Override
    public void resume() {
        Global.lifecycle.resume();
    }

    @Override
    public void dispose() {
        Global.lifecycle.dispose();
    }

}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/Game.kt
package com.binarymonks.jj.core

import com.badlogic.gdx.ApplicationListener
import com.binarymonks.jj.core.specs.SceneSpec

/**
 * A Game
 *
 * You must extend this and only ever create one of them. It is the entry point to Jenjin and it initializes
 * a bunch of global state - which must only be done once.
 *
 * @property jjConfig Configuration for your game.
 */
abstract class Game (val jjConfig: JJConfig = JJConfig()) : ApplicationListener{

    override fun create() {
        JJ.initialise(this.jjConfig)
        gameOn()
    }

    override fun pause() {
    }

    override fun resize(width: Int, height: Int) {
    }
    override fun render() {
        JJ.B.time.update()
//        JJ.assets.update()
//        JJ.B.tasks.preloopTasks.update()
//        JJ.B.thingWorld.update()
        JJ.B.layers.update()
        JJ.B.physicsWorld.update()
//        JJ.B.tasks.postPhysicsTasks.update()
    }

    override fun resume() {
    }

    override fun dispose() {
    }

    /**
     * This is your entry hook.
     *
     * You can instantiate assets and build your [SceneSpec]s here or whatever.
     * In the end though, you need to probably do 1 of 3 things:
     *
     *  - Load a [SceneSpec] or
     *  - Present a Splash Screen or
     *  - Present a Menu
     *
     */
    protected abstract fun gameOn()

}


